src/bin/pg_autoctl/ipaddr.c

```diff
@@ -244,3 +302,226 @@ isTestEnv(void)
+	 */
+	for (dns_addr=dns_lookup_addr; dns_addr; dns_addr = dns_addr->ai_next)
+	{
+		/*
+		 * Skip loopback addresses, we want to be able to connect to
+		 * hostname from other nodes.
+		 */
+		if (!(dns_addr->ai_flags & IFF_LOOPBACK))
+		{
+			for (ifa = ifaddr; ifa; ifa = ifa->ifa_next)
+			{
+				/*
+				 * TODO: implement ipv6 support
+				 */
+				if (ifa->ifa_addr->sa_family == AF_INET
+					&& dns_addr->ai_family == AF_INET)
+				{
+					struct sockaddr_in *ip =
+						(struct sockaddr_in *) ifa->ifa_addr;
+
+					if (ipv4eq(ip, (struct sockaddr_in *) dns_addr->ai_addr))
+					{
+						/*
+						 * Found an IP address in the DNS answer that
+						 * matches one of the interfaces IP addresses on
+						 * the machine.
+						 */
+						freeaddrinfo(dns_lookup_addr);
+
+						if (inet_ntop(AF_INET,
+									  (void *) &(ip->sin_addr),
+									  localIpAddress,
+									  size) == NULL)
+						{
+							log_warn("Failed to determine local ip address: "
+									  "%s", strerror(errno));
+							return false;
+						}
+
+						return true;
+					}
+				}
+				else if (ifa->ifa_addr->sa_family == AF_INET6
+						 && dns_addr->ai_family == AF_INET6)
+				{
+					struct sockaddr_in6 *ip =
+						(struct sockaddr_in6 *) ifa->ifa_addr;
+
+					if (ipv6eq(ip, (struct sockaddr_in6 *) dns_addr->ai_addr))
+					{
+						/*
+						 * Found an IP address in the DNS answer that
+						 * matches one of the interfaces IP addresses on
+						 * the machine.
+						 */
+						freeaddrinfo(dns_lookup_addr);
+
+						if (inet_ntop(AF_INET6,
+									  (void *) &(ip->sin6_addr),
+									  localIpAddress,
+									  size) == NULL)
+						{
+							/* check size >= INET6_ADDRSTRLEN */
+							log_warn("Failed to determine local ip address: "
+									  "%s", strerror(errno));
+							return false;
+						}
+
+						return true;
+					}
+				}
+			}
+		}
+	}
+
+	freeaddrinfo(dns_lookup_addr);
+	return false;
+}
+
+
+/*
+ * ip_address_type parses the hostname and determines whether it is an IPv4
+ * address, IPv6 address, or DNS name.
+ *
+ * To edit pg HBA file, when given an IP address (rather than a hostname), we
+ * need to compute the CIDR mask. In the case of ipv4, that's /32, in the case
+ * of ipv6, that's /128. The `ip_address_type' function discovers which type of
+ * IP address we are dealing with.
+ */
+IPType
+ip_address_type(const char *hostname)
+{
+	struct in_addr ipv4;
+	struct in6_addr ipv6;
+
+	if (hostname == NULL)
+	{
+		return IPTYPE_NONE;
+	}
+	else if (inet_pton(AF_INET, hostname, &ipv4) == 1)
+	{
+		log_trace("hostname \"%s\" is ipv4", hostname);
+		return IPTYPE_V4;
+	}
+	else if (inet_pton(AF_INET6, hostname, &ipv6) == 1)
+	{
+		log_trace("hostname \"%s\" is ipv6", hostname);
+		return IPTYPE_V6;
+	}
+	return IPTYPE_NONE;
+}
+
+
+/*
+ * findHostnameFromLocalIpAddress does a reverse DNS lookup from the local IP
+ * address that we found in fetchLocalIPAddress, given as a parameter.
```

Comment:
the comment is a bit specific about how to get to the `localIpAddress`, wonder if we could word it a bit more generic as I think it is already used from multiple locations

Reply:
Improved comment.